### PR TITLE
Add Whitelist Support

### PR DESCRIPTION
**Pull request by joerchan**
Follow up on whitelisting support API proposal from: 
https://github.com/zephyrproject-rtos/zephyr/pull/12874#discussion_r254601048

Add whitelist support in the bluetooth host.
Supported features:
 - Advertising with whitelist on scan requests, connect request ,or both
 - Scanning with whitelist
 - Creating connections using a whitelist (Auto connection procedure).

### CODE REVIEW

**Comment by Vudentz** [comment]
I mostly fine with these new APIs, Im just wondering if the controller don't have enough entries if it wouldn't be possible to have the host implementing it own whitelist as well, or do some sort of rotation of the entries on controller, though perhaps that is left for the board/platform to have enough entries to fulfill the use case but in case of host-only build it may not be possible to customize the number of entries.

**Comment by joerchan** [comment]
> I mostly fine with these new APIs, Im just wondering if the controller don't have enough entries if it wouldn't be possible to have the host implementing it own whitelist as well, or do some sort of rotation of the entries on controller, though perhaps that is left for the board/platform to have enough entries to fulfill the use case but in case of host-only build it may not be possible to customize the number of entries.

Doing whitelisting in the host is not really possible, because whitelisting happens without Host involvement at all. 
So for example an advertiser would have to allow all connections, and then immediately disconnect unwanted peers.
Restricting scan responses would not be possible at all.
Initiator could not be used, so we would be back again to using scanner, so really no improvement there.

File: include/bluetooth/bluetooth.h
Comment (carlescufi) [inline include/bluetooth/bluetooth.h]: `FILTER_SCAN_REQ`

File: include/bluetooth/bluetooth.h
Comment (carlescufi) [inline include/bluetooth/bluetooth.h]: Update the doxygen for `filter_dup`

File: include/bluetooth/bluetooth.h
Comment (carlescufi) [inline include/bluetooth/bluetooth.h]: identity?

File: include/bluetooth/bluetooth.h
Comment (carlescufi) [inline include/bluetooth/bluetooth.h]: identity?

**Review by carlescufi — CHANGES_REQUESTED** [review]
A few naming comments, but otherwise looks pretty good

File: include/bluetooth/bluetooth.h
Comment (MarekPieta) [inline include/bluetooth/bluetooth.h]: missing space before */

File: include/bluetooth/bluetooth.h
Comment (MarekPieta) [inline include/bluetooth/bluetooth.h]: `whitelist`?
use when or if, not both :)

File: include/bluetooth/bluetooth.h
Comment (MarekPieta) [inline include/bluetooth/bluetooth.h]: same as above

File: include/bluetooth/bluetooth.h
Comment (MarekPieta) [inline include/bluetooth/bluetooth.h]: same as above

File: subsys/bluetooth/host/Kconfig
Comment (MarekPieta) [inline subsys/bluetooth/host/Kconfig]: Do we need help here (it provides no new information) ?

File: subsys/bluetooth/host/conn.c
Comment (MarekPieta) [inline subsys/bluetooth/host/conn.c]: remove additional empty line

File: subsys/bluetooth/shell/bt.c
Comment (MarekPieta) [inline subsys/bluetooth/shell/bt.c]: `Clearing whitelist failed ?`

**Review by MarekPieta — COMMENTED** [review]
Please also fix space before comma in commit: Bluetooth: Host: Add whitelist support in Bluetooth Host API

File: include/bluetooth/conn.h
Comment (pdunaj) [inline include/bluetooth/conn.h]: So will it work similarly to direct advertising?
i.e. we would set whitelist and call `bt_conn_create_auto_le` instead of `bt_le_adv_start`?

File: subsys/bluetooth/host/conn.c
Comment (pdunaj) [inline subsys/bluetooth/host/conn.c]: Maybe these functions could do:
```
if (!IS_ENABLED(CONFIG_BT_WHITELIST)) {
   return -ENOTSUP;
}
```
instead of using preprocessor. As this is not enabled by default it would be safer to let it compiler (and check code) every time.

File: subsys/bluetooth/host/hci_core.c
Comment (pdunaj) [inline subsys/bluetooth/host/hci_core.c]: why cast to `void *` instead of `struct bt_hci_rp_le_read_wl_size *`?

File: subsys/bluetooth/host/hci_core.h
Comment (pdunaj) [inline subsys/bluetooth/host/hci_core.h]: dot missing at the end. also in other places

File: include/bluetooth/bluetooth.h
Comment (joerchan) [reply include/bluetooth/bluetooth.h]: Yes, that sentence got a bit messed up. :)

File: subsys/bluetooth/host/Kconfig
Comment (joerchan) [reply subsys/bluetooth/host/Kconfig]: Added some more information.

File: subsys/bluetooth/host/conn.c
Comment (joerchan) [reply subsys/bluetooth/host/conn.c]: @pdunaj Since the variables in `bt_dev` are compiled out to save memory. This code cannot be compiled without using preprocessor.
The test application for bluetooth shell enables the BT_WHITELIST, so it will be compiled as part of CI.

File: include/bluetooth/conn.h
Comment (joerchan) [reply include/bluetooth/conn.h]: Well no, because directed advertising is with `bt_conn_create_slave_le`, and that is only towards one peer.
It will work the same as with `bt_le_adv_start` and `option = BT_LE_ADV_OPT_FILTER_CONN`, then all devices in the whitelist are able to connect.

File: subsys/bluetooth/host/hci_core.c
Comment (joerchan) [reply subsys/bluetooth/host/hci_core.c]: Absolutely no reason (I just duplicated the rl_size read), will change :)

File: include/bluetooth/bluetooth.h
Comment (joerchan) [reply include/bluetooth/bluetooth.h]: Updated

**Review by carlescufi — APPROVED** [review]
LGTM now, but @jhedberg should take a look before we merge

File: subsys/bluetooth/host/conn.c
Comment (pdunaj) [reply subsys/bluetooth/host/conn.c]: ok

**Comment by ioannisg** [comment]
> 
> 
> LGTM now, but @jhedberg should take a look before we merge

@jhedberg is it possible that you can look at it before 2.0 feature freeze?

File: subsys/bluetooth/host/hci_core.c
Comment (Vudentz) [inline subsys/bluetooth/host/hci_core.c]: Im not a found of declaration in the middle of the function, specially since this is a conditional block, we can probably have a cleaner code by moving this into its own function.

File: subsys/bluetooth/shell/bt.c
Comment (MarekPieta) [inline subsys/bluetooth/shell/bt.c]: Do we need this assignment ?

**Review by MarekPieta — APPROVED** [review]

File: subsys/bluetooth/host/hci_core.c
Comment (joerchan) [reply subsys/bluetooth/host/hci_core.c]: I will create an _complete function, which seems to be the pattern.

File: include/bluetooth/bluetooth.h
Comment (joerchan) [reply include/bluetooth/bluetooth.h]: Not without rewriting it.

File: subsys/bluetooth/shell/bt.c
Comment (joerchan) [reply subsys/bluetooth/shell/bt.c]: No, removed.

**Review by Vudentz — APPROVED** [review]
LGTM

**Pull request merged**
